### PR TITLE
Remove unnecessary calls to float() before division.

### PR DESCRIPTION
**Pull request by anntzer**
Most of these calls probably date back to the time before
`__future__`-division existed...

Note that this will have the side effect that some functions that were
accidentally accepting strings instead of floats before will no longer
accept strings -- which seems fine to me.

## PR Checklist

- [ ] Has Pytest style unit tests
- [ ] Code is PEP 8 compliant
- [ ] New features are documented, with examples if plot related
- [ ] Documentation is sphinx and numpydoc compliant
- [ ] Added an entry to doc/users/next_whats_new/ if major new feature (follow instructions in README.rst there)
- [ ] Documented in doc/api/api_changes.rst if API changed in a backward-incompatible way

### CODE REVIEW

File: lib/matplotlib/testing/compare.py [hunk 1]
```diff
@@ -367,12 +367,7 @@ def calculate_rms(expectedImage, actualImage):
         raise ImageComparisonFailure(
             "Image sizes do not match expected size: {0} "
             "actual size {1}".format(expectedImage.shape, actualImage.shape))
-    num_values = expectedImage.size
-    abs_diff_image = abs(expectedImage - actualImage)
-    histogram = np.bincount(abs_diff_image.ravel(), minlength=256)
-    sum_of_squares = np.sum(histogram * np.arange(len(histogram)) ** 2)
-    rms = np.sqrt(float(sum_of_squares) / num_values)
-    return rms
+    return ((expectedImage - actualImage) ** 2).mean()
```
Comment (afvincent) [inline lib/matplotlib/testing/compare.py]: Huh, I may have missed something but this is "Mean Square" rather than "Root Mean Square", is it not? Is it not missing a call to `np.sqrt`?

File: lib/mpl_toolkits/mplot3d/proj3d.py [hunk 2]
```diff
@@ -61,7 +61,7 @@ def line2d_seg_dist(p1, p2, p0):
     x01 = np.asarray(p0[0]) - p1[0]
     y01 = np.asarray(p0[1]) - p1[1]
 
-    u = (x01*x21 + y01*y21)/float(abs(x21**2 + y21**2))
+    u = (x01*x21 + y01*y21) / abs(x21**2 + y21**2)
```
Comment (afvincent) [inline lib/mpl_toolkits/mplot3d/proj3d.py]: Very genuine comment: are `x21` and `y21` complex or real numbers? Because in the latter case, what is the point of `abs(...)` then?

File: examples/color/color_cycle_default.py [hunk 3]
```diff
@@ -13,7 +13,7 @@
 colors = prop_cycle.by_key()['color']
 
 lwbase = plt.rcParams['lines.linewidth']
-thin = float('%.1f' % (lwbase / 2))
+thin = lwbase / 2
```
Comment (afvincent) [inline examples/color/color_cycle_default.py]: By curiosity, was there a purpose to the former "rounding to 0.1" scheme?

**Comment by afvincent** [comment]
I have the feeling that Travis agrees with me about the "typo" in the image comparison test 😈.

File: lib/matplotlib/testing/compare.py (hunk 1 above)
Comment (tacaswell) [reply lib/matplotlib/testing/compare.py]: The previous method (going through the histogram) seems extra complex, why was it done that way?  Is this just a weird historical artifact of supporting numarray/numeric or an attempt to get performance?

File: lib/mpl_toolkits/axisartist/angle_helper.py [hunk 4]
```diff
@@ -56,10 +52,10 @@ def select_step_hour(dv):
     minsec_limits_ = [1.5, 2.5, 3.5, 4.5, 5.5, 8, 11, 14, 18, 25, 45]
     minsec_steps_  = [1,   2,   3,   4,   5,   6, 10, 12, 15, 20, 30]
 
-    minute_limits_ = A(minsec_limits_)*(1./60.)
+    minute_limits_ = np.array(minsec_limits_) / 60
```
Comment (tacaswell) [inline lib/mpl_toolkits/axisartist/angle_helper.py]: This looks like another case of old-school micro-optomizations like (`x*x` vs `x**2`), but this one still work!

```
In [3]: d = np.array(range(52))

In [4]: %timeit d / 60
936 ns ± 0.538 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [5]: %timeit d * (1/ 60)
683 ns ± 1.43 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

In [6]: d = np.array(range(50000))

In [7]: %timeit d / 60
102 µs ± 95.3 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [8]: %timeit d * (1/ 60)
37.4 µs ± 554 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)


```

**Comment by tacaswell** [comment]
:+1: in principle, but some concern about the RMS changes.

File: lib/matplotlib/testing/compare.py (hunk 1 above)
Comment (anntzer) [reply lib/matplotlib/testing/compare.py]: the missing sqrt is just a typo, fixed.
pretty sure performance of this specific part of the code is irrelevant compared to rendering the png or shelling out to inkscape and ghostscript...

**Comment by anntzer** [comment]
hopefully fixed the rms changes

File: lib/mpl_toolkits/axisartist/angle_helper.py (hunk 4 above)
Comment (anntzer) [reply lib/mpl_toolkits/axisartist/angle_helper.py]: I can repro, but I think axisartist/angle_helper is far enough from being on a critical code path to make it worth obfuscating...

Comment (tacaswell) [reply lib/mpl_toolkits/axisartist/angle_helper.py]: :+1:

File: lib/mpl_toolkits/mplot3d/proj3d.py (hunk 2 above)
Comment (anntzer) [reply lib/mpl_toolkits/mplot3d/proj3d.py]: Pretty sure they're intended to be reals (it's computing a 3D distance...).  Removed the call to abs, good catch.

File: examples/color/color_cycle_default.py (hunk 3 above)
Comment (anntzer) [reply examples/color/color_cycle_default.py]: Probably to make the title label look better(?), but I think 0.75 looks just fine too.

Comment (afvincent) [reply examples/color/color_cycle_default.py]: Fair enough, and I think the same

File: examples/api/mathtext_asarray.py [hunk 5]
```diff
@@ -22,7 +22,7 @@
     r'some other string', color='red', fontsize=20, dpi=200)
 
 fig = plt.figure()
-fig.figimage(rgba1.astype(float)/255., 100, 100)
-fig.figimage(rgba2.astype(float)/255., 100, 300)
+fig.figimage(rgba1, 100, 100)
```
Comment (efiring) [inline examples/api/mathtext_asarray.py]: Why did you remove the division by 255?

Comment (anntzer) [reply examples/api/mathtext_asarray.py]: because we can display uint8 images just as well as float images, and the real point of the example is to show how to get the mathtext as an image array...

File: lib/matplotlib/backends/backend_mixed.py [hunk 6]
```diff
@@ -139,8 +139,8 @@ def stop_rasterizing(self):
                 #       backends support this.
                 self._renderer.draw_image(
                     gc,
-                    float(l) / self.dpi * self._figdpi,
-                    (float(height)-b-h) / self.dpi * self._figdpi,
+                    l / self.dpi * self._figdpi,
+                    (height-b-h) / self.dpi * self._figdpi,
```
Comment (efiring) [inline lib/matplotlib/backends/backend_mixed.py]: To me, since you are in here, these would be much clearer as
```python
self._figdpi / self.dpi,
self._figdpi * (height-b-h) / self.dpi,
```
Seeing `a/b*c` always slows me down in a way that `c*a/b` or `(a/b)*c` doesn't.

Comment (anntzer) [reply lib/matplotlib/backends/backend_mixed.py]: agreed, fixed

File: lib/mpl_toolkits/axisartist/angle_helper.py [hunk 7]
```diff
@@ -250,7 +245,7 @@ def _get_number_fraction(self, factor):
                 break
 
             d = factor // threshold
-            int_log_d = int(floor(math.log10(d)))
+            int_log_d = int(math.floor(math.log10(d)))
```
Comment (efiring) [inline lib/mpl_toolkits/axisartist/angle_helper.py]: Earlier it looked like you were eliminating the use of the math module in favor of numpy functions--but not here. Why?  (The math module is faster; I assumed you were simplifying by eliminating it in contexts where the speed difference doesn't matter.)

File: lib/matplotlib/testing/compare.py [hunk 8]
```diff
@@ -367,12 +367,8 @@ def calculate_rms(expectedImage, actualImage):
...
             "Image sizes do not match expected size: {0} "
             "actual size {1}".format(expectedImage.shape, actualImage.shape))
-    num_values = expectedImage.size
-    abs_diff_image = abs(expectedImage - actualImage)
-    histogram = np.bincount(abs_diff_image.ravel(), minlength=256)
-    sum_of_squares = np.sum(histogram * np.arange(len(histogram)) ** 2)
-    rms = np.sqrt(float(sum_of_squares) / num_values)
-    return rms
+    # Convert to float to avoid overflowing finite integer types.
+    return np.sqrt(((expectedImage - actualImage).astype(float) ** 2).mean())
```
Comment (efiring) [inline lib/matplotlib/testing/compare.py]: This is seems to be a substantive change to a genuine RMS, from something I don't understand.  There must have been a reason for the convoluted original.

File: lib/matplotlib/ticker.py [hunk 9]
```diff
@@ -1574,7 +1574,7 @@ def tick_values(self, vmin, vmax):
         """
         if self.nbins is None:
             return self.locs
-        step = max(int(0.99 + len(self.locs) / float(self.nbins)), 1)
+        step = max(int(np.ceil(len(self.locs) / self.nbins)), 1)
```
Comment (efiring) [inline lib/matplotlib/ticker.py]: Won't this change the results in rare cases?  It does look nicer, though.

File: lib/mpl_toolkits/axes_grid1/axes_size.py [hunk 10]
```diff
@@ -268,7 +268,7 @@ def from_any(size, fraction_ref=None):
         return Fixed(size)
     elif isinstance(size, six.string_types):
         if size[-1] == "%":
-            return Fraction(float(size[:-1])/100., fraction_ref)
+            return Fraction(float(size[:-1]) / 100, fraction_ref)
```
[the remark this answers is not in the document]
Comment (anntzer) [reply lib/mpl_toolkits/axes_grid1/axes_size.py]: because the purpose here is to convert `"3%`' to `0.03`? :)

File: lib/matplotlib/testing/compare.py (hunk 8 above)
Comment (anntzer) [reply lib/matplotlib/testing/compare.py]: I don't understand the original either, but the relevant tests are still passing...

File: lib/mpl_toolkits/axisartist/angle_helper.py (hunk 7 above)
Comment (anntzer) [reply lib/mpl_toolkits/axisartist/angle_helper.py]: because the rest of the expression used math and I was too lazy to change both.  Obviously this is not on a critical path, changed.

File: lib/matplotlib/ticker.py (hunk 9 above)
Comment (anntzer) [reply lib/matplotlib/ticker.py]: 1) I think the code really, really looks like it was written by someone who meant ceil but didn't remember that function existed.
2) For the change to actually matter you need a) to use bins with FixedLocator, which seems to be quite rare in my experience, and b) at least 99 ticks and 100 bins (to fall in the "edge" region), which is also quite rare...

File: lib/mpl_toolkits/axes_grid1/axes_size.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (efiring) [reply lib/mpl_toolkits/axes_grid1/axes_size.py]: Right.  Sorry, I didn't pay enough attention to the context.

File: lib/matplotlib/testing/compare.py (hunk 8 above)
Comment (efiring) [reply lib/matplotlib/testing/compare.py]: Ours are, but I suspect this might cause problems for packagers who put in larger tolerances than ours, as I believe some do.  Therefore I recommend sticking to the original algorithm for the purposes of this PR.  Perhaps add a comment that this is *not* a simple rms.  @mdboom can probably tell us the rationale for the algorithm.

File: lib/matplotlib/ticker.py (hunk 9 above)
Comment (efiring) [reply lib/matplotlib/ticker.py]: I'll go along with this one.

File: lib/matplotlib/testing/compare.py (hunk 8 above)
Comment (anntzer) [reply lib/matplotlib/testing/compare.py]: AFAICT the results should be the same: the old method counted, for each delta, the number of pixels that have this delta value, then computed a RMS based on that histogram.
Again, https://github.com/matplotlib/matplotlib/pull/10335/files shows tests where we check that the reported RMS values are stable to 1e-4.

Comment (efiring) [reply lib/matplotlib/testing/compare.py]: If I am correctly expressing the two algorithms below, then their results are not the same:
```
In [26]: x = np.zeros(1000, dtype=int); x[0] = 1

In [27]: h = np.bincount(x, minlength=256)

In [28]: np.sqrt(((h * np.arange(len(h)))**2).mean())
0.0625

In [29]: np.sqrt((x**2).mean())
0.031622776601683791

In [30]: x[:10]
array([1, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [31]: (h * np.arange(len(h)))[:10]
array([0, 1, 0, 0, 0, 0, 0, 0, 0, 0])

In [32]: len(x)
1000

In [33]: len(h)
256
```

Comment (anntzer) [reply lib/matplotlib/testing/compare.py]: the previous algorithm was dividing by num_values (so sum() / len(x) in your case, not mean()).  this fix shows that the values are indentical.

Comment (efiring) [reply lib/matplotlib/testing/compare.py]: Oops... sorry I missed that.

Comment (QuLogic) [reply lib/matplotlib/testing/compare.py]: Well, I can run the Cartopy test suite without any issue using this PR.
